### PR TITLE
Enable debugging on Reanimated's runtime using Chrome DevTools

### PR DESCRIPTION
**Pull request by Kwasow**
## Description

This PR is related to #1960 (and its replacement #2047). It adds the ability to use Chrome DevTools to add breakpoints and debug worklets (or the UI context in general) both on Android and iOS.

## Major changes

Runtime creation has been moved to `ReanimatedRuntime` for both Android and iOS (in the `Common/cpp/ReanimatedRuntime` directory).

Before (Android) [file: `NativeProxy.cpp`]:
```cpp
#if JS_RUNTIME_HERMES
  auto config =
      ::hermes::vm::RuntimeConfig::Builder().withEnableSampleProfiling(false);
  std::shared_ptr<jsi::Runtime> animatedRuntime =
      facebook::hermes::makeHermesRuntime(config.build());
#elif JS_RUNTIME_V8
      …
#else
      …
#endif
```

After (shared) [file: `NativeProxy.cpp`]:
```cpp
ReanimatedRuntime::make(jsQueue);
```

The custom build config has been removed, as sample profiling is set to false by default.

The created `HermesRuntimeManager` object is the stored inside `ReanimatedNativeModule` as it is important that it’s lifetime is synced with the lifetime of the module.

## Testing

- [x] Builds on Android
- [x] Builds on Android in release mode
- [x] App reloads work on Android
- [x] Builds on iOS
- [x] Builds on iOS in release mode
- [x] App reloads work on iOS
- [x] JSC still works
- [x] JSC debugging in Safari with iOS still works
- [x] Paper still works

Versions of React-Native tested: 0.70

## Things to look into

- ~~Breakpoint labels do not appear on iOS~~
_I tested this on a new app with Flipper on the main JS thread, and it was also the case, so it seem to not be an issue on our side_
- ~~After removing a breakpoint on iOS it can't be set in the same location again~~
_This seems to be an issue with Chrome DevTools as connecting to Reanimated's runtime through Flipper fully works_
- ~~Edge-case: the app will crash if a reload is performed while the debugger is open~~
_Will be fixed in a PR to metro and [58e9e7b](https://github.com/software-mansion/react-native-reanimated/pull/3526/commits/58e9e7bcd7d3b6b590a5f6fca0db93a951eaa39e)_
- The latest release of Chrome (105) broke source maps, so only Flipper works now. I reported the issue [here](https://bugs.chromium.org/p/chromium/issues/detail?id=1360298#makechanges), and I'm waiting for a response from the Chromium team

## TODO

- [x] Open PR to update debugging docs in documentation (will be included in #3446)
- [x] Open PR in [facebook/flipper](https://github.com/facebook/flipper) to enable Flipper support on custom runtimes (facebook/flipper#4047)
- [x] Open PR in [facebook/metro](https://github.com/facebook/metro) to support debugger reloads on custom runtimes (facebook/metro#864)

### CODE REVIEW

File: Common/cpp/ReanimatedRuntime/ReanimatedRuntime.cpp [hunk 1]
```diff
@@ -0,0 +1,45 @@
...
+      facebook::hermes::makeHermesRuntime();
+  facebook::hermes::HermesRuntime &hermesRuntime = *runtime;
+
+  return std::make_shared<ReanimatedHermesRuntime>(
+      std::move(runtime), hermesRuntime, jsQueue);
+#elif JS_RUNTIME_V8
+  jsQueue->quitSynchronous();
+
+  auto config = std::make_unique<rnv8::V8RuntimeConfig>();
+  config->enableInspector = false;
```
Comment (Kudo) [inline Common/cpp/ReanimatedRuntime/ReanimatedRuntime.cpp:35]: fwiw, if you want to enable debugging on v8, you could also set this flag as true.

File: Common/cpp/ReanimatedRuntime/ReanimatedRuntime.cpp [hunk 2]
```diff
@@ -0,0 +1,45 @@
...
+  config->appName = "reanimated";
+  return rnv8::createSharedV8Runtime(runtime_, std::move(config));
+#else
+  jsQueue->quitSynchronous();
```
Comment (tomekzaw) [inline Common/cpp/ReanimatedRuntime/ReanimatedRuntime.cpp:41]: Why is `quitSynchronous` call present only for V8 and JSC?

File: android/src/fabric/java/com/swmansion/reanimated/NativeProxy.java [hunk 3]
```diff
@@ -294,8 +297,8 @@ public void prepare(LayoutAnimations LayoutAnimations) {
     }
     mNodesManager = mContext.get().getNativeModule(ReanimatedModule.class).getNodesManager();
     FabricUIManager fabricUIManager =
-        (FabricUIManager) UIManagerHelper.getUIManager(mContext.get(), UIManagerType.FABRIC);
-    installJSIBindings(fabricUIManager);
+      (FabricUIManager) UIManagerHelper.getUIManager(mContext.get(), UIManagerType.FABRIC);
+    installJSIBindings(new ReanimatedMessageQueueThread(), fabricUIManager);
```
Comment (tomekzaw) [inline android/src/fabric/java/com/swmansion/reanimated/NativeProxy.java:301]: Let's construct this object in the previous line, i.e.
```java
MessageQueueThread reanimatedMessageQueueThread = new ReanimatedMessageQueueThread();
```

File: Common/cpp/ReanimatedRuntime/ReanimatedHermesRuntime.cpp [hunk 4]
```diff
@@ -0,0 +1,86 @@
+#include "ReanimatedHermesRuntime.h"
```
Comment (tomekzaw) [inline Common/cpp/ReanimatedRuntime/ReanimatedHermesRuntime.cpp:1]: You can move this to `#if JS_RUNTIME_HERMES` if you want.

Btw why don't we compile this file only `#if JS_RUNTIME_HERMES && HERMES_ENABLE_DEBUGGER`?

File: Common/cpp/ReanimatedRuntime/ReanimatedHermesRuntime.h [hunk 5]
```diff
@@ -0,0 +1,125 @@
...
+#include <cxxreact/MessageQueueThread.h>
+#include <jsi/decorator.h>
+#include <jsi/jsi.h>
+
+#include <memory>
+#include <thread>
+
+#if __has_include(<reacthermes/HermesExecutorFactory.h>)
+#include <reacthermes/HermesExecutorFactory.h>
+#else // __has_include(<hermes/hermes.h>) or ANDROID
```
Comment (tomekzaw) [inline Common/cpp/ReanimatedRuntime/ReanimatedHermesRuntime.h:23]:
```suggestion
#else // __has_include(<hermes/hermes.h>) || ANDROID
```

File: Common/cpp/ReanimatedRuntime/ReanimatedRuntime.h [hunk 6]
```diff
@@ -0,0 +1,26 @@
+#pragma once
+
+// JS_RUNTIME_HERMES is only set on Android so we have to check __has_include
+// on iOS.
+#if (__has_include( \
+         <reacthermes/HermesExecutorFactory.h>) || __has_include(<hermes/hermes.h>) || JS_RUNTIME_HERMES)
+#define JS_RUNTIME_HERMES 1
+#endif
```
Comment (tomekzaw) [inline Common/cpp/ReanimatedRuntime/ReanimatedRuntime.h:8]: Same as above

File: Common/cpp/ReanimatedRuntime/RuntimeInitialization.md [hunk 7]
```diff
@@ -0,0 +1,194 @@
...
+The initialization process is pretty simple and has only been moved out of
+`NativeProxy` into `ReanimatedRuntime` without any major changes.
+
+## Hermes runtime debugging
+
+To enable debugging on the Hermes runtime we need to do two things:
+
+1. Include source maps in JavaScript files
+
+This part is done purely in JavaScript via the babel plugin. The `makeWorklet`
```
Comment (tomekzaw) [inline Common/cpp/ReanimatedRuntime/RuntimeInitialization.md:25]:
```suggestion
This part is done purely in JavaScript via the Babel plugin. The `makeWorklet`
```

File: RNReanimated.podspec [hunk 8]
```diff
@@ -104,9 +110,9 @@ Pod::Spec.new do |s|
     "HEADER_SEARCH_PATHS" => "\"$(PODS_TARGET_SRCROOT)/ReactCommon\" \"$(PODS_TARGET_SRCROOT)\" \"$(PODS_ROOT)/#{folly_prefix}Folly\" \"$(PODS_ROOT)/boost\" \"$(PODS_ROOT)/boost-for-react-native\" \"$(PODS_ROOT)/DoubleConversion\" \"$(PODS_ROOT)/Headers/Private/React-Core\"",
     "CLANG_CXX_LANGUAGE_STANDARD" => "c++17",
   }
-  s.compiler_flags = folly_compiler_flags + ' ' + boost_compiler_flags
+  s.compiler_flags = folly_compiler_flags + ' ' + boost_compiler_flags + ' -DHERMES_ENABLE_DEBUGGER'
```
Comment (tomekzaw) [inline RNReanimated.podspec:113]: Do we really want to set `-DHERMES_ENABLE_DEBUGGER` even in release mode?

**Review by tomekzaw — CHANGES_REQUESTED** [review]

File: ios/native/NativeProxy.mm [hunk 9]
```diff
@@ -30,13 +30,8 @@
...
 
-#if __has_include(<reacthermes/HermesExecutorFactory.h>)
-#import <reacthermes/HermesExecutorFactory.h>
-#elif __has_include(<hermes/hermes.h>)
-#import <hermes/hermes.h>
-#else
-#import <jsi/JSCRuntime.h>
-#endif
+#import "REAMessageThread.h"
+#import "ReanimatedRuntime.h"
```
[the remark this answers is not in the document]
Comment (Kwasow) [reply ios/native/NativeProxy.mm:34]: I also moved the imports up to where other `RNReanimated` imports are

File: RNReanimated.podspec (hunk 8 above)
Comment (Kwasow) [reply RNReanimated.podspec:113]: This is the same as in React Native https://github.com/facebook/react-native/blob/68f3a42fc7380051714253f43b42175de361f8bd/ReactCommon/hermes/React-hermes.podspec#L40

File: Common/cpp/ReanimatedRuntime/ReanimatedRuntime.cpp [hunk 10]
```diff
@@ -0,0 +1,45 @@
...
+  return std::make_shared<ReanimatedHermesRuntime>(
+      std::move(runtime), hermesRuntime, jsQueue);
+#elif JS_RUNTIME_V8
+  jsQueue->quitSynchronous();
+
+  auto config = std::make_unique<rnv8::V8RuntimeConfig>();
+  config->enableInspector = false;
+  config->appName = "reanimated";
+  return rnv8::createSharedV8Runtime(runtime_, std::move(config));
+#else
```
[the remark this answers is not in the document]
Comment (Kwasow) [reply Common/cpp/ReanimatedRuntime/ReanimatedRuntime.cpp:38]: Let's do it in a different PR and also look through the entire codebase to make the appropriate changes

File: Common/cpp/ReanimatedRuntime/ReanimatedHermesRuntime.cpp (hunk 4 above)
Comment (Kwasow) [reply Common/cpp/ReanimatedRuntime/ReanimatedHermesRuntime.cpp:1]: We could initialize the runtime differently, when `HERMES_ENABLE_DEBUGGER` is not set, but I think it looks cleaner when it is consistent. Though I could wrap the adapter in `#if HERMES_ENABLE_DEBUGGER` to exclude it in non-debug builds

File: plugin.js [hunk 11]
```diff
@@ -358,8 +402,13 @@ function makeWorklet(t, fun, state) {
...
   // some even break them
 
+  const codeObject = generate(fun.node, {
+    sourceMaps: true,
+    sourceFileName: state.file.opts.filename,
+  });
+
   const code =
-    '\n(' + (t.isObjectMethod(fun) ? 'function ' : '') + fun.toString() + '\n)';
+    '(' + (t.isObjectMethod(fun) ? 'function ' : '') + codeObject.code + '\n)';
```
[the remark this answers is not in the document]
Comment (Kwasow) [reply plugin.js:411]: It is, I added a comment

**Comment by tomekzaw** [comment]
https://user-images.githubusercontent.com/20516055/191472289-c56fe037-9efe-4289-ae9c-632573e82c7c.mp4

**Review by tomekzaw — APPROVED** [review]

**Comment by tomekzaw** [comment]
https://user-images.githubusercontent.com/20516055/191474676-b56e940c-0ffd-4c79-8ff7-1b5ba19a782a.mp4

**Comment by tomekzaw** [comment]
Congratulations on this PR! You have completed a task that many have attempted but no one succeeded. A lot of research had to be done to make this happen. Certainly, it takes a lot of skills and patience to dive into and navigate through the deepest trenches of React Native ecosystem, in particular exploring undocumented areas of codebase of external tools such as Flipper and Metro. The solution you have proposed involves several programming languages including C++, Objective-C, Ruby, Java, Groovy, TypeScript and Flow as well as various development tools such as npm, yarn, Babel, Gradle, CMake, CocoaPods and Docusaurus. It was a pleasure to watch the progress of the work, follow your findings and conclusions, and finally discuss various aspects of the implementation, from code style to architecture and versioning. This feature is a game-changer and it will certainly be highly appreciated by the community. Thank you!

File: plugin.js [hunk 12]
```diff
@@ -418,17 +469,6 @@ function makeWorklet(t, fun, state) {
...
-      // test for <something>.value = <something> expressions
-      const left = path.node.left;
-      if (
-        t.isMemberExpression(left) &&
-        t.isIdentifier(left.object) &&
-        t.isIdentifier(left.property, { name: 'value' })
-      ) {
-        outputs.add(left.object.name);
-      }
-    },
```
[the remark this answers is not in the document]
Comment (Kwasow) [reply plugin.js:431]: In the function, we were creating `outputs = new Set()`. This piece of code adds some things to the set, but we never use that set afterwards. It is an old piece of code that used to do something, but after some changes it became useless and forgotten, which is the fate that awaits us all.

**Review by piaskowyk — APPROVED** [review]
Well done!

**Comment by capezzbr** [comment]
This PR broke integration with V8 engine

```
node_modules/react-native-reanimated/Common/cpp/ReanimatedRuntime/ReanimatedRuntime.cpp:40:38: error: use of undeclared identifier 'runtime_'
```

**Comment by tomekzaw** [comment]
Thanks for letting us know @capezzbr, we'll definitely need to take a look at this. We don't have any V8 example app in our repo yet so there's no CI that checks V8 integration. I will try to setup the app anytime soon.

**Comment by capezzbr** [comment]
Thank you as always @tomekzaw 🙏

**Comment by capezzbr** [comment]
Hey @tomekzaw, here is my attempt to fix the issue: https://github.com/software-mansion/react-native-reanimated/pull/3736. Let me know what you think 👍

**Pull request merged**